service/src/main/java/bio/terra/workspace/service/danglingresource/DanglingResourceCleanupService.java

```diff
@@ -0,0 +1,213 @@
+          configuration.getPollingInterval().toSeconds(),
+          TimeUnit.SECONDS);
+    }
+  }
+
+  /**
+   * Run {@code cleanupResources}, suppressing all thrown exceptions. This is helpful as {@code
+   * ScheduledExecutorService.scheduleAtFixedRate} will stop running if any execution throws an
+   * exception. Suppressing these exceptions ensures we do not stop cleaning up resources if a
+   * single run fails.
+   */
+  public void cleanupResourcesSuppressExceptions() {
+    try {
+      cleanupResources();
+    } catch (Exception e) {
+      LoggingUtils.logAlert(
+          logger,
+          "Unexpected error during danglingResourceCleanup execution, see stacktrace below");
+      logger.error("danglingResourceCleanup stacktrace: ", e);
+    }
+  }
+
+  private void cleanupResources() {
+    if (!configuration.isEnabled()) {
+      return;
+    }
+    logger.info("Beginning dangling resource cleanup cronjob");
+    // Use a one-second shorter duration here to ensure we don't skip a run by moving slightly too
+    // quickly.
+    Duration claimTime = configuration.getPollingInterval().minus(Duration.ofSeconds(1));
+    // Attempt to claim the latest run of this job to ensure only one pod runs the cleanup job.
+    if (!cronjobDao.claimJob(DANGLING_RESOURCE_CLEANUP_JOB_NAME, claimTime)) {
+      logger.info("Another pod has executed this job more recently. Ending resource cleanup.");
+      return;
+    }
+
+    // Read all resource entries that are in the READY state and match the resource type of
+    // potential dangling resources.
+    List<ControlledResource> resourcesToCheck =
+        resourceDao.listControlledResourcesByType(DANGLING_RESOURCE_TYPES);
+    logger.info("Checking {} resources to see if they are dangling.", resourcesToCheck.size());
+
+    String wsmSaToken = samService.getWsmServiceAccountToken();
+    AuthenticatedUserRequest wsmSaRequest =
+        new AuthenticatedUserRequest().token(Optional.of(wsmSaToken));
+
+    // For each resource, check if it exists in the cloud. If it does not, run the cleanup flight to
+    // delete its db metadata entry and associated sam resource.
+    for (ControlledResource resource : resourcesToCheck) {
+      if (!cloudResourceExists(resource)) {
+        logger.info(
+            "Cleaning up dangling resource {} of type {}.",
+            resource.getResourceId(),
+            resource.getResourceType());
+        launchDanglingResourceCleanupFlight(resource, wsmSaRequest);
+      }
+    }
+  }
+
+  /**
+   * Launches a resource delete flight for a given dangling resource. The controlled resource
+   * deletion flight is reused here to delete the sam resource and metadata entry.
+   */
+  private void launchDanglingResourceCleanupFlight(
+      ControlledResource resource, AuthenticatedUserRequest wsmSaRequest) {
+    String jobId = UUID.randomUUID().toString();
+    controlledResourceService.deleteControlledResourceAsync(
```

Comment:
this flight DeleteControlledResourcesFlight can actually takes in multiple resources. so instead of launching one flight per resources ( i think there's a limit on how many flights there can be atm), can you pass in all the resources at once?
the input param is CONTROLLED_RESOURCES_TO_DELETE.

Also if you specify operationType.system_cleanup, it will be system cleanup. currently it's just delete.

Reply:
My initial thinking was that by using individual flights per resource deletion, we can provide flight metadata such as workspaceId + resourceId/type, etc. In fact, it seems like workspaceId is a requirement to execute the flight, so I'm not sure how we would run a flight to delete resources across multiple workspaces.

Thoughts on still cleaning up resources in individual flights but updating the operation type to SYSTEM_CLEANUP?

Reply:
ooh i forgot about that.. sure